### ISSUE TITLE
New release

### ISSUE THREAD

**Issue by anmonteiro** [ocaml-ppx/ppx_deriving_yojson#83]
This library hasn't seen a release in a while. Would it be possible to release to OPAM?

I can help if needed

**Comment by gasche** [comment]
Yes, of course it is reasonable to do an OPAM release. Your help would be appreciated because right now I am too short on time to do it myself.

How to build, test and release `ppx_deriving_yojson` is completely flushed out of my short-term memory, especially since I'm not yet familiar with dune-specific release automation ( ppx_deriving_yojson recently migrated to dune, #69 ). Help for any of the following would be immensely helpful:

- Test ppx_deriving_yojson on your end, and report here the build/test commands that you used.
- The build files may need updating; I see that the dune workspace does not have a 4.07 context, adding one would help.
- The dependencies in the `.opam` file have no version constraints, this sounds a bit too simple to be true. We need to test that ppx_deriving_yojson still builds with the most recent versions of its dependencies (or add high bounds to restrict to older version), and possibly add lower-bound versions. (There was one for ppx_import in the opam repository, but it looks like this testing dependency has vanished?)
- The [Changelog](https://github.com/ocaml-ppx/ppx_deriving_yojson/blob/master/CHANGELOG.md) needs updating with a list of user-visible changes that happened since the 3.1 release.

Feel free to send a PR for any subset of these tasks, it is very helpful. When all of that is done, I can easily add and push a 3.2 tag (I'm not sure if tags can be merged by PRs?). Then a pointer to the proper package-upload workflow for dune projects would again save me some time, but I can also look myself.

**Comment by anmonteiro** [comment]
@gasche I opened https://github.com/ocaml-ppx/ppx_deriving_yojson/pull/85 with most of your suggestions, and some additional work. I explain them in more detail below.

> * Test ppx_deriving_yojson on your end, and report here the build/test commands that you used.

I tested `ppx_deriving_yojson` on my end, both by running the tests and in a project that uses it (pinned the development version). All is good. `make` / `make build` and `make test` work fine.

Furthermore, I included a `CONTRIBUTING.md` file that documents how to set up and run all these for future reference.

> * The build files may need updating; I see that the dune workspace does not have a 4.07 context, adding one would help.

I ported this project from `jbuilder` to `dune` and added a 4.07 context.

I also updated the OPAM files to v2, as well as the CI scripts.

> * The dependencies in the `.opam` file have no version constraints, this sounds a bit too simple to be true. We need to test that ppx_deriving_yojson still builds with the most recent versions of its dependencies (or add high bounds to restrict to older version), and possibly add lower-bound versions. (There was one for ppx_import in the opam repository, but it looks like this testing dependency has vanished?)

This actually worked for me in my project that uses a pinned version of `ppx_deriving_yojson` so I think we don't actually have to do anything here?

> * The [Changelog](https://github.com/ocaml-ppx/ppx_deriving_yojson/blob/master/CHANGELOG.md) needs updating with a list of user-visible changes that happened since the 3.1 release.

This is not yet in the PR, but I'll be adding them shortly.

> Feel free to send a PR for any subset of these tasks, it is very helpful. When all of that is done, I can easily add and push a 3.2 tag (I'm not sure if tags can be merged by PRs?). Then a pointer to the proper package-upload workflow for dune projects would again save me some time, but I can also look myself.

Unfortunately I don't think tags can be merged from PRs.

I think I've done everything you suggested but adding the CHANGELOG (which I'll do shortly) and instructions on how to cut a release. It's unclear to me how to publish to OPAM 2 (they're not accepting 1.x releases anymore) and the current `opam-query` setup doesn't seem to support OPAM 2. Something to investigate further, I suppose.

**Comment by gasche** [comment]
Thanks a lot! This is great.

> This actually worked for me in my project that uses a pinned version of ppx_deriving_yojson so I think we don't actually have to do anything here?

Sounds good. Let's leave it at that for now; the OPAM repository has continuous integration that will catch breakage with recent versions of our dependencies, so if you overlooked something we should still notice.

> [Changelog entries are] not yet in the PR, but I'll be adding them shortly.

Great!

> Unfortunately I don't think tags can be merged from PRs.

I'll do it on my end.

> It's unclear to me how to publish to OPAM 2 (they're not accepting 1.x releases anymore) and the current opam-query setup doesn't seem to support OPAM 2.

I think that cool kids use [dune-release](https://github.com/samoht/dune-release) these days, but I am familiar with the manual workflow of creating an opam-repository PR, so I will do this part on my end.

**Comment by anmonteiro** [comment]
I just updated the CHANGELOG in #85. Apart from the instructions on how to do a manual release, it should be good to go?

**Comment by gasche** [comment]
I tagged the release, and sent a PR to the opam repository: https://github.com/ocaml/opam-repository/pull/13146

**Comment by gasche** [comment]
So https://github.com/ocaml-ppx/ppx_deriving_yojson/issues/76 causes compatibility. I'm pausing the release and looking at this right now. @anmonteiro, if you happen to have some time for it, I might invite you to review code and test a fix on your end.

**Comment by anmonteiro** [comment]
@gasche definitely. Ping me in the PR and I'll do it

**Comment by gasche** [comment]
New opam-repository PR at

  https://github.com/ocaml/opam-repository/pull/13150

Let's see what the revdeps testing says...

**Comment by gasche** [comment]
The new release is now released on opam, closing.

Thanks a lot @anmonteiro for your gentle push and your help, and @emillon and @whitequark for your short-order review and comments.
